src/compiler/checker.ts

```diff
@@ -22826,39 +22852,35 @@ namespace ts {
 
                 // If the symbol of the node has members, treat it like a constructor.
                 const symbol = getSymbolOfNode(func);
-                return !!symbol && (symbol.members !== undefined || symbol.exports !== undefined && symbol.exports.get("prototype" as __String) !== undefined);
-            }
-            return false;
-        }
-
-        function isJSConstructorType(type: Type) {
-            if (type.flags & TypeFlags.Object) {
-                const resolved = resolveStructuredTypeMembers(<ObjectType>type);
-                return resolved.callSignatures.length === 1 && isJSConstructor(resolved.callSignatures[0].declaration);
+                return !!symbol && symbol.members !== undefined;
             }
             return false;
         }
 
-        function getJSClassType(symbol: Symbol): Type | undefined {
-            let inferred: Type | undefined;
-            if (isJSConstructor(symbol.valueDeclaration)) {
-                inferred = getInferredClassType(symbol);
+        function mergeJSSymbols(target: Symbol, source: Symbol | undefined) {

```

Comment:
If `mergeJSSymbols` is called twice with the same pair of symbols, you'll get a new output symbol - I think the result of the merge needs to be stored, eg, on the symbol links of the target.

Reply:
It's not future-proof, but we both looked at the callers of mergeJSSymbols and they both cache immediately after calling it.